blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/BlueOceanCredentialsProvider.java

```diff
@@ -0,0 +1,262 @@
+import com.cloudbees.plugins.credentials.CredentialsStore;
+import com.cloudbees.plugins.credentials.CredentialsStoreAction;
+import com.cloudbees.plugins.credentials.common.IdCredentials;
+import com.cloudbees.plugins.credentials.domains.Domain;
+import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import com.google.common.collect.ImmutableSet;
+import hudson.Extension;
+import hudson.model.ItemGroup;
+import hudson.model.ModelObject;
+import hudson.model.TopLevelItem;
+import hudson.model.User;
+import hudson.security.Permission;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.acegisecurity.Authentication;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link CredentialsProvider} to serve credentials stored in user store.
+ *
+ * It works by looking for {@link FolderPropertyImpl} on a folder. This should allow
+ * using such user scoped properties with {@link jenkins.branch.OrganizationFolder} or {@link jenkins.branch.MultiBranchProject}
+ * or even a Folder.
+ *
+ * @author Stephen Connoly
+ * @author Vivek Pandey
+ */
+@Extension(optional = true)
+public class BlueOceanCredentialsProvider extends CredentialsProvider {
+    private static final String DISPLAY_NAME = "BlueOcean Folder Credentials";
+
+    @Nonnull
+    @Override
+    public <C extends Credentials> List<C> getCredentials(@Nonnull Class<C> type, @Nullable ItemGroup itemGroup,
+                                                          @Nullable Authentication authentication) {
+        if (itemGroup instanceof AbstractFolder) {
+            return getCredentials(type, (AbstractFolder)itemGroup);
+        }
+        return Collections.emptyList();
+    }
+
+    private static <C extends Credentials> List<C> getCredentials(Class<C> type, AbstractFolder folder){
+        FolderPropertyImpl prop = (FolderPropertyImpl)folder.getProperties().get(FolderPropertyImpl.class);
+        if (prop != null) {
+            User user = User.get(prop.getUser(), false, Collections.emptyMap());
+            if(user != null){
+                Authentication auth = user.impersonate();
+                return CredentialsMatchers.filter(
+                        CredentialsProvider.lookupCredentials(
+                                type,
+                                Jenkins.getInstance(),
+                                auth,
+                                Collections.<DomainRequirement>emptyList()
+                        ),
+                        CredentialsMatchers.allOf(CredentialsMatchers.withId(prop.getId()),
+                                CredentialsMatchers.withScope(CredentialsScope.USER))
```

Comment:
I am now wondering if perhaps we shouldn't be more explicit and just target the credential stores scoped to the user and bypass the general `lookupCredentials`, e.g.

```
for (CredentialStore store: CredentialsProvider.lookupStores(user)) {
  // if store has credentials with id, return credential
}
// return empty list;
```

as matching based on scope here is ringing some alarm bells